### DISCUSSION TITLE
[AuthSession] How to use with Splitwise API

### DISCUSSION THREAD

**Discussion by airen29** [expo/expo#9597] [Archived: Questions] [answered]
Hi guys. I have an issue that I am not able to solve anyhow.

I am trying to get an `code` from `Splitwise API` that is successful. Then I want to recieve `accessToken` and I am getting only `{}` as an answer:

Code
```javascript
import React, { useEffect } from 'react';
import { Button, View, Text } from 'react-native';
import * as AuthSession from 'expo-auth-session';
import { useAuthRequest } from 'expo-auth-session';
import { maybeCompleteAuthSession } from 'expo-web-browser';

maybeCompleteAuthSession();

const ReceiptDetailsScreen = (props) => {
  const { navigation } = props;
  const { route: { params: fakeData } } = props;

  const redirectUri = AuthSession.makeRedirectUri({
    path: 'redirect',
    preferLocalhost: true,
    native: 'exp://localhost:19000',
  });
  
  let discovery = {
    authorizationEndpoint: 'https://secure.splitwise.com/oauth/authorize',
    tokenEndpoint: 'https://secure.splitwise.com/oauth/token',
  }
  let config = {
    clientId: 'MY-CLIENT-ID',
    clientSecret: 'MY-CLIENT-SECRET',
    redirectUri,
    scopes: [],
  };

  const [request, response, promptAsync] = useAuthRequest(config, discovery);

  useEffect(() => {
    if (response?.type === 'success') {
      console.log(response);
      console.log(request);
      }
  }, [response]);

  const getToken = async () => {
    try {
      console.log(response.params.code);

      const accessToken = new AuthSession.AccessTokenRequest({
        clientId: config.clientId,
        clientSecret: config.clientSecret,
        code: response.params.code,
        redirectUri: config.redirectUri,
        scopes: config.scopes,
      }, { tokenEndpoint: discovery.tokenEndpoint });

      console.log(accessToken);

      const responseToken = await AuthSession.exchangeCodeAsync(accessToken, discovery);
      console.log(JSON.stringify(responseToken, null, 4))
    } catch (err) {
      console.log(err);
      console.log(JSON.stringify(err, null, 4))
    };
  };

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
        <Text>Receipt Details Screen</Text>
        { fakeData.receipt.items.map(item => (<Text>{ item.name } </Text>))}

        <Button
          onPress={() => promptAsync({ redirectUri })}
          title="Splitwise account"
        />

        {response && <Button title="Obtain Token" onPress={() => getToken()} />}
    </View>
    );
}

export default ReceiptDetailsScreen;
```

Response in console is:
```
MY-OBTAINED-CODE
AccessTokenRequest {
  "clientId": "MY-CLIENT-ID",
  "clientSecret": "MY-CLIENT-SECRET",
  "code": "MY-OBTAINED-CODE",
  "extraParams": undefined,
  "grantType": "authorization_code",
  "redirectUri": "exp://localhost:19000/--/redirect",
  "request": Object {
    "clientId": "MY-CLIENT-ID",
    "clientSecret": "MY-CLIENT-SECRET",
    "code": "MY-OBTAINED-CODE",
    "redirectUri": "exp://localhost:19000/--/redirect",
    "scopes": Array [],
  },
  "scopes": Array [],
}
{}

JSON Parse error: Unexpected EOF
* [native code]:null in parse
- node_modules/promise/setimmediate/core.js:37:13 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:24 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:135:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:183:16 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:446:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:396:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:144:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:373:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:4 in flushedQueue
* [native code]:null in flushedQueue
* [native code]:null in invokeCallbackAndReturnFlushedQueue
```

Any information what I am doing wrong? After clicking on button `Obtain Token` my application returns an error in the console without `accessToken` as an answer or any `result`.

Thanks a lot for any help. Maybe I am using `API` wrong.

**Comment by EvanBacon** [comment, marked as answer]
Can confirm splitwise works with expo-auth-session.

## Setup

You will need a new app for each platform as Splitwise only supports a single callback URI (and doesn't seem to use the `redirect_uri` query parameter). This setup is for an expo web project (in development) started using https `expo web --https`:

<img width="1437" alt="Screen Shot 2020-08-06 at 1 46 48 PM" src="https://user-images.githubusercontent.com/9664363/89581145-56b1ca00-d7eb-11ea-8607-614b3125d043.png">

You'll need a new provider app for each platform you plan on using due to the single callback URL restriction on Splitwise.

<img width="1437" alt="Screen Shot 2020-08-06 at 2 33 05 PM" src="https://user-images.githubusercontent.com/9664363/89584976-dfcbff80-d7f1-11ea-88b2-e2ba7471ad40.png">

## Code

Be sure to use the _implicit_ auth flow with `ResponseType.Token`. Don't use the `clientSecret` in the application code, if you need it, then setup a serverless type function using netlify or now. You shouldn't need it for this though.

```ts
const redirectUri = AuthSession.makeRedirectUri({
    native: 'bareexpo://redirect',
    path: 'redirect',
    preferLocalhost: true,
});

const [request, result, promptAsync] = AuthSession.useAuthRequest(
    {
      clientId: 'CLIENT_ID',
      // Use response_type=token instead to get the access_token returned in the initial request.
      responseType: AuthSession.ResponseType.Token,
      redirectUri,
      scopes: [],
    },
    // discovery
    {
      authorizationEndpoint: 'https://secure.splitwise.com/oauth/authorize',
      tokenEndpoint: 'https://secure.splitwise.com/oauth/token',
    }
);

/** 
 * Later:
 * `result.params === { access_token: "ewz******************b", state "D**********a" }`
 */

// I recommend the following window popup size for desktop browsers as it makes the splitwise content render correctly:
promptAsync({ /* useProxy goes here if you use it */,  windowFeatures: { width: 960, height: 500 } })
```

## Demo

<img width="1072" alt="Screen Shot 2020-08-06 at 1 50 24 PM" src="https://user-images.githubusercontent.com/9664363/89581593-1e5ebb80-d7ec-11ea-837c-3272e1598165.png">

**Reply by airen29** [reply]
Thank you very much. I managed to make it work as you mentioned. I also managed to make it work via Postman so I think that with classic style (what I was trying) it would be possible too. Problem is in 99 % callback URI.
They do use `redirect_uri` parameter, but all the content must be in the `body`, not in the `url` as params. This was my mistake.

I will again try to make it work via `iOS device` on the phone (this is again I think only a problem with wrong callback URI).

Thanks again for help. Maybe later we could create official Expo Splitwise API tutorial.

**Comment by darryl1975** [comment]
Hi guys,

I tried the following code to login to Azure AD and am successfuly to login to Azure.

let client_Id = 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx';
  let client_Secret = 'yyyyyyyyyyyy_yyyyyyy-yyy-yyyyyyyyy';
  let scopesList: string[] = ['openid', 'profile', 'email', 'offline_access'];

  let authorization_Endpoint = 'https://login.microsoftonline.com/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx/oauth2/authorize';
  let token_Endpoint = 'https://login.microsoftonline.com/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx/oauth2/token';

  // AuthRequestConfig
  let reqConfig: AuthSession.AuthRequestConfig = {
    //responseType: ResponseType.Token,
    clientId: client_Id,
    clientSecret: client_Secret,
    redirectUri: redirect_uri,
    scopes: scopesList,
    usePKCE: true
  }

  let discoveryConfig: AuthSession.DiscoveryDocument = {
    authorizationEndpoint: authorization_Endpoint,
    tokenEndpoint: token_Endpoint
  }

  let authRequestConfig: AuthSession.AuthRequestConfig = {
    clientId: client_Id,
    clientSecret: client_Secret,
    redirectUri: redirect_uri,
    scopes: scopesList,
    prompt: AuthSession.Prompt.Login
  };

  // Create and load an auth request
  const [request, result, promptAsync] = AuthSession.useAuthRequest(
    reqConfig,
    discoveryConfig
  );


However,  the response doesn't provide me the email or id_token or access_token.

I tried to use the return code provided in response.params.code to retrieve a token. However, I am unsuccessful.

The following is the code I have to retrieve a token (on button clicked):

const getToken = async () => {
    try {
      var rslt: any = result;
      console.log('App getToken (rslt): ', rslt.params.code);

      const accessToken = new AuthSession.AccessTokenRequest({
        clientId: client_Id,
        clientSecret: client_Secret,
        code: rslt.params.code,
        redirectUri: redirect_uri,
        scopes: scopesList,
      });

      console.log(accessToken);

      const responseToken = await AuthSession.exchangeCodeAsync(accessToken, { tokenEndpoint: discovery.tokenEndpoint });
      console.log(JSON.stringify(responseToken, null, 4))
    } catch (err) {
      console.log(err);
      console.log(JSON.stringify(err, null, 4))
    };
  };


The error I encountered:

App retrieveToken (error):  [Error: The provided authorization grant (e.g., authorization code, resource owner credentials) or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.
More info: AADSTS50148: The code_verifier does not match the code_challenge supplied in the authorization request for PKCE.
Trace ID: 256b595a-2503-4d44-83e4-1ac91add3400
Correlation ID: e74b0bd0-52ef-4f65-923b-2c0afee976e1
Timestamp: 2020-10-29 03:38:28Z]
Unrecognized event: {"type":"client_log","level":"log","data":["App retrieveToken (error): ","[Error: The provided authorization grant (e.g., authorization code, resource owner credentials) or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.\nMore info: AADSTS50148: The code_verifier does not match the code_challenge supplied in the authorization request for PKCE.\r\nTrace ID: 256b595a-2503-4d44-83e4-1ac91add3400\r\nCorrelation ID: e74b0bd0-52ef-4f65-923b-2c0afee976e1\r\nTimestamp: 2020-10-29 03:38:28Z]"]}
